src/devices/machine/i82586.cpp

```diff
@@ -73,47 +160,1692 @@ device_memory_interface::space_config_vector i82586_base_device::memory_space_co
 
 WRITE_LINE_MEMBER(i82586_base_device::ca)
 {
-	logerror("%s: channel attention %s (%s)\n", state ? "asserted" : "deasserted", tag(), machine().describe_context());
+	LOG("channel attention %s (%s)\n", state ? "asserted" : "deasserted", machine().describe_context());
+
+	if (state)
+	{
+		// on first ca after reset, initialise
+		if (!m_initialised)
+			initialise();
+		else
+			process_scb();
+	}
 }
 
-void i82586_device::device_start()
+void i82586_base_device::recv_cb(u8 *buf, int length)
 {
-	i82586_base_device::device_start();
+	switch (m_ru_state)
+	{
+	case RU_IDLE:
+	case RU_SUSPENDED:
+		// discard all frames
+		break;
+
+	case RU_READY:
+		if (accept_filter(buf))
+		{
+			LOG("recv_cb receiving frame length %d\n", length);
+			dump_bytes(buf, length);
+
+			ru_execute(buf, length);
+
+			update_scb();
+		}
+		break;
+
+	default:
+		// no resources
+		// TODO: accumulate statistics
+		break;
+	}
 }
 
-void i82586_device::device_reset()
+void i82586_base_device::process_scb()
 {
-	i82586_base_device::device_reset();
+	// fetch current command and status
+	m_scb_cs = m_space->read_dword(m_scb_address);
+
+#if VERBOSE & LOG_GENERAL
+	static const char *cuc[] = { "NOP", "START", "RESUME", "SUSPEND", "ABORT", "THROTTLE_D", "THROTTLE_I", "reserved" };
```

Comment:
These and others like them should be `const char *const whatever[]` or the pointers are not constant, only the characters they point to.